source/simulator.hpp

```diff
@@ -10,42 +15,174 @@
+    }
+
+    std::shared_ptr<TReceiver> add_receiver(std::string name) {
+        if (m_receivers.contains(name)) {
+            LOG_WARN(fmt::format(
+                "add_receiver failed: device with name {} already exists.",
+                name));
+            return nullptr;
+        }
+        m_receivers[name] = std::make_shared<TReceiver>();
+        return m_receivers[name];
+    }
+
+    std::shared_ptr<TSwitch> add_switch(std::string name) {
+        if (m_switches.contains(name)) {
+            LOG_WARN(fmt::format(
+                "add_switch failed: device with name {} already exists.",
+                name));
+            return nullptr;
+        }
+        m_switches[name] = std::make_shared<TSwitch>();
+        return m_switches[name];
+    }
 
-    std::shared_ptr<Flow> add_flow(std::shared_ptr<ISender> sender,
-                                   std::shared_ptr<IReceiver> recever,
-                                   Size a_packet_size,
-                                   Time a_delay_between_packets,
-                                   std::uint32_t packets_to_send);
+    std::shared_ptr<TFlow> add_flow(std::shared_ptr<ISender> sender,
+                                    std::shared_ptr<IReceiver> receiver,
+                                    Size packet_size,
+                                    Time delay_between_packets,
+                                    std::uint32_t packets_to_send) {
+        auto flow =
+            std::make_shared<Flow>(sender, receiver, packet_size,
+                                   delay_between_packets, packets_to_send);
+        m_flows.emplace_back(flow);
+        return flow;
+    }
 
     void add_link(std::shared_ptr<IRoutingDevice> a_from,
                   std::shared_ptr<IRoutingDevice> a_to,
-                  std::uint32_t a_speed_mbps, Time a_delay);
+                  std::uint32_t a_speed_mbps, Time a_delay) {
+        auto link = std::make_shared<Link>(a_from, a_to, a_speed_mbps, a_delay);
+        m_links.emplace_back(link);
+        a_from->add_outlink(link);
+        a_to->add_inlink(link);
+    }
 
-    std::vector<std::shared_ptr<IRoutingDevice>> get_devices() const;
+    std::vector<std::shared_ptr<IRoutingDevice>> get_devices() const {
+        std::vector<std::shared_ptr<IRoutingDevice>> result;
+        std::transform(m_senders.begin(), m_senders.end(),
+                       std::back_inserter(result),
+                       [](const auto& pair) { return pair.second; });
+
+        std::transform(m_receivers.begin(), m_receivers.end(),
+                       std::back_inserter(result),
+                       [](const auto& pair) { return pair.second; });
+
+        std::transform(m_switches.begin(), m_switches.end(),
+                       std::back_inserter(result),
+                       [](const auto& pair) { return pair.second; });
+        return result;
+    }
 
     // Clear all events in the Scheduler
-    void clear();
+    void clear() {}
```

Comment:
Please, move realisation to cpp and add clearing scheduler:
```suggestion
    void clear() {
        Scheduler::get_instance()->clear();
    }
```